### PR TITLE
C#, replace the current Xform method with a * operator.

### PR DESCRIPTION
**Pull request by magian1127**
Test on 3.4
GDS supports Xform(Rect2) and Xform(PoolVector2Array).
Now add the implementation of C#.

测试于3.4
GDS是支持 Xform(Rect2) 和 Xform(PoolVector2Array).
现在添加了C#的实现.

### CODE REVIEW

**Comment by aaronfranke** [comment]
@magian1127 Thank you for working on this. However, in the current master branch, the Xform method has been replaced by the `*` operator in GDScript. https://github.com/godotengine/godot/blob/master/doc/classes/Transform2D.xml#L128

There are three things that should be done and you are welcome to help with them (all in C#):

* In master, replace the current Xform method with a `*` operator.
* In master, add new `*` operators for Rect2 and PackedVector2Array.
* Backport the `*` operators to Godot 3.x (this will need a dedicated PR due to the PoolVector2Array name).

**Comment by magian1127** [comment]
@aaronfranke 

I revised it to see if it meets the requirements. I passed the test myself.
xform_ Inv remains the same, right?

我修改了好了, 看下是否符合,我自己测试通过.
xform_inv依旧保持不变是吧?

File: modules/mono/glue/GodotSharp/GodotSharp/Core/Transform2D.cs [hunk 1]
```diff
@@ -355,11 +355,45 @@ public Transform2D Translated(Vector2 offset)
...
             return new Vector2(Tdotx(v), Tdoty(v)) + origin;
         }
 
+        /// <summary>
+        /// Returns a Rect2 transformed (multiplied) by this transformation matrix.
+        /// </summary>
+        /// <param name="r">A Rect2 to transform.</param>
+        /// <returns>The transformed Rect2.</returns>
+        [Obsolete("Xform is deprecated. Use the multiplied(Transform2D * Rect2) instead.")]
+        public Rect2 Xform(Rect2 r)
```
Comment (aaronfranke) [inline modules/mono/glue/GodotSharp/GodotSharp/Core/Transform2D.cs:370]: We don't need to add a method that is immediately obsolete. We only need the operators.

File: modules/mono/glue/GodotSharp/GodotSharp/Core/Vector2.cs [hunk 2]
```diff
@@ -645,6 +645,17 @@ public Vector2(Vector2 v)
...
 
+        /// <summary>
+        /// Returns a Vector2 transformed (multiplied) by transformation matrix.
+        /// </summary>
+        /// <param name="t">transformation</param>
+        /// <param name="v">A Vector2 to transform.</param>
+        /// <returns>The transformed Vector2.</returns>
+        public static Vector2 operator *(Transform2D t, Vector2 v)
+        {
+            return v * t;
```
Comment (aaronfranke) [inline modules/mono/glue/GodotSharp/GodotSharp/Core/Vector2.cs:656]: This is not correct. Matrix multiplication is not commutative. `transform * vector` is the same as `transform.xform(vector)`, and `vector * transform` is the same as `transform.xform_inv(vector)`. See these lines in `core/variant/variant_op.cpp` for reference:

```cpp
register_op<OperatorEvaluatorXForm<Vector2, Transform2D, Vector2>>(Variant::OP_MULTIPLY, Variant::TRANSFORM2D, Variant::VECTOR2);
register_op<OperatorEvaluatorXFormInv<Vector2, Vector2, Transform2D>>(Variant::OP_MULTIPLY, Variant::VECTOR2, Variant::TRANSFORM2D);
```

Comment (magian1127) [reply modules/mono/glue/GodotSharp/GodotSharp/Core/Vector2.cs]: I don't have enough knowledge. I'm sorry

I'll know where to look next time there are similar changes

File: modules/mono/glue/GodotSharp/GodotSharp/Core/Transform2D.cs (hunk 1 above)
Comment (neikeq) [reply modules/mono/glue/GodotSharp/GodotSharp/Core/Transform2D.cs]: Yes, if this is for master (4.0-dev) then just remove the method. For 3.x the obsolete attribute is good.

**Review by neikeq — APPROVED** [review]

File: modules/mono/glue/GodotSharp/GodotSharp/Core/Transform2D.cs [hunk 3]
```diff
@@ -365,6 +366,7 @@ public Vector2 Xform(Vector2 v)
         /// </summary>
         /// <param name="v">A vector to inversely transform.</param>
         /// <returns>The inversely transformed vector.</returns>
+        [Obsolete("XformInv is deprecated. Use the multiplied(Vector2 * Transform2D) instead.")]
```
Comment (akien-mga) [inline modules/mono/glue/GodotSharp/GodotSharp/Core/Transform2D.cs:369]:
```suggestion
        [Obsolete("XformInv is deprecated. Use the multiplication operator (Vector2 * Transform2D) instead.")]
```

File: modules/mono/glue/GodotSharp/GodotSharp/Core/Transform2D.cs [hunk 4]
```diff
@@ -355,6 +355,7 @@ public Transform2D Translated(Vector2 offset)
         /// </summary>
         /// <param name="v">A vector to transform.</param>
         /// <returns>The transformed vector.</returns>
+        [Obsolete("Xform is deprecated. Use the multiplied(Transform2D * Vector2) instead.")]
```
Comment (akien-mga) [inline modules/mono/glue/GodotSharp/GodotSharp/Core/Transform2D.cs:358]:
```suggestion
        [Obsolete("Xform is deprecated. Use the multiplication operator (Transform2D * Vector2) instead.")]
```

**Comment by akien-mga** [comment]
Could you squash the commits together? See [PR workflow](https://docs.godotengine.org/en/latest/community/contributing/pr_workflow.html#the-interactive-rebase) for instructions.

**Comment by magian1127** [comment]
Ok... I'll operate it now
好的.. 我现在去操作.

**Review by akien-mga — APPROVED** [review]

**Comment by akien-mga** [comment]
Thanks!

**Pull request merged**